### ISSUE TITLE
Improve IovReader/IovWriter achitecture

### ISSUE THREAD

**Issue by webmaster128** [iov-one/iov-core#291]
## Dependencies

The currently documented way to create an IovReader is complicated:

```js
const knownChains = await withConnectors([await bnsConnector("ws://localhost:22345")]);
// reader is already there: knownChains[0].client, no writer required

const writer = new IovWriter(profile, knownChains);
const chainId = writer.chainIds()[0];
const reader = writer.reader(chainId);
```

So that the moment we need a reader to make a writer but need a writer to make a reader.

## IovReader is responsible for writing to the blockchain

The term "reader" is irritating as IovReader is not only reading from the blockchain but also write signed transactions to it. It serves more as a networking component.

This becomes important when thinking about `disconnect()`. Right now, in order to disconnect a writer, you need to disconnect all the readers and as soon as you disconnect a reader, the writer does not work anymore. This behavior would make more sense, if it was clear, that IovReader is IovWriter's networking component.

**Comment by ethanfrey** [comment]
The uglyness commented on above was fixed with #323 
The naming clarity part is still outstanding.

I would suggest `IovReader` -> `ChainConnection` or `BcpConnection` or something similar

`IovWriter` is a multi-chain signer. If we look at the separation of components in #301, we consider when this is out of process (eg. in extension, mobile phone), and the client using the `IovWriter` would maintain it's own set of `Connections` for queries.

I would also suggest naming the concrete class `IovWriter` -> `MultiChainSigner` or such, and defining an interface (with `chainId()`, but without `getReader()` functionality) that can refer to the in-process or out of process key management and transaction signing portion.

Maybe `Wallet`?

In short, my suggestion:

* `IovReader` interface -> `BcpConnection` (or similar)
* `IovWriter` concrete class -> `MultiChainSigner` (with a disconnect that handles all connections)
* `MultiChainSigner` implementes new`Wallet` interface, which we use in all the signing code. This `Wallet` does the signing can asks for user-approval, designed for the case where the `TransactionRequest` and the private key management are performed by different processes.

**Comment by webmaster128** [comment]
From my point of view, this issue is mostly done. The current usage is much simpler and consistent

```ts
const writer = new IovWriter(profile);
await writer.addChain(bnsConnector("ws://localhost:22345"));
const chainId = writer.chainIds()[0];
const connection = writer.reader(chainId); // a BCP connection
```

I'd just rename the getter `reader(chainId: ChainId): BcpConnection` to `connection(chainId: ChainId): BcpConnection` before closing.

**Comment by ethanfrey** [comment]
I agree on that rename.
I would also like to rename `IovWriter` to something nicer.
Since `Wallet` is taken, I gotta dream again...

Authenticator? Signer? Bank? 

Please give me a good name. This is something we will also want to connect with out-of-process to handle private keys and signing material. In that case, we won't expose the connection() over the wire.

I mean I think `IovWriter` should implement an `interface` that explains the above concept. And then have a name to match the interface name (like `Local*`, `Embedded*`, etc.)

**Comment by webmaster128** [comment]
I can live with IovWriter for now until we find a natural fit. It is unique and not misleading. From the ideas that exist now, I like MultiChainSigner most.

**Comment by Isabello** [comment]
`MultiChainSigner` seems pretty good and explicit about what it does. I like it.

**Comment by ethanfrey** [comment]
This can be closed now, only remaining task is it's own issue #416
